Index: code/ColladaExporter.cpp
===================================================================
--- code/ColladaExporter.cpp	(リビジョン 1347)
+++ code/ColladaExporter.cpp	(作業コピー)
@@ -124,8 +124,10 @@
 	mOutput << startstr << "</contributor>" << endstr;
   mOutput << startstr << "<created>2000-01-01T23:59:59</created>" << endstr;
   mOutput << startstr << "<modified>2000-01-01T23:59:59</modified>" << endstr;
-	mOutput << startstr << "<unit name=\"centimeter\" meter=\"0.01\" />" << endstr;
-	mOutput << startstr << "<up_axis>Y_UP</up_axis>" << endstr;
+	//mOutput << startstr << "<unit name=\"centimeter\" meter=\"0.01\" />" << endstr;
+	mOutput << startstr << "<unit name=\"meter\" meter=\"1.0\" />" << endstr;
+	//mOutput << startstr << "<up_axis>Y_UP</up_axis>" << endstr;
+	mOutput << startstr << "<up_axis>Z_UP</up_axis>" << endstr;
 	PopTag();
 	mOutput << startstr << "</asset>" << endstr;
 }
Index: code/ColladaLoader.cpp
===================================================================
--- code/ColladaLoader.cpp	(リビジョン 1347)
+++ code/ColladaLoader.cpp	(作業コピー)
@@ -146,7 +146,7 @@
 
 	// ... then fill the materials with the now adjusted settings
 	FillMaterials(parser, pScene);
-
+#if 0
 	// Convert to Y_UP, if different orientation
 	if( parser.mUpDirection == ColladaParser::UP_X)
 		pScene->mRootNode->mTransformation *= aiMatrix4x4( 
@@ -160,6 +160,20 @@
 			 0,  0,  1,  0,
 			 0, -1,  0,  0,
 			 0,  0,  0,  1);
+#endif
+	// Convert to Z_UP, if different orientation
+	if( parser.mUpDirection == ColladaParser::UP_X)
+		pScene->mRootNode->mTransformation *= aiMatrix4x4(
+			 0,  0, -1,  0,
+			 0,  1,  0,  0,
+			 1,  0,  0,  0,
+			 0,  0,  0,  1);
+	else if( parser.mUpDirection == ColladaParser::UP_Y)
+		pScene->mRootNode->mTransformation *= aiMatrix4x4(
+			 1,  0,  0,  0,
+			 0,  0, -1,  0,
+			 0,  1,  0,  0,
+			 0,  0,  0,  1);
 
 	// store all meshes
 	StoreSceneMeshes( pScene);
