diff --git code/ColladaExporter.cpp code/ColladaExporter.cpp
index 7634806..a961c92 100644
--- code/ColladaExporter.cpp
+++ code/ColladaExporter.cpp
@@ -292,7 +293,8 @@ void ColladaExporter::WriteMaterials()
       WriteTextureParamEntry( mat.specular, "specular", mat.name);
       WriteTextureParamEntry( mat.reflective, "reflective", mat.name);
 
-      mOutput << startstr << "<technique sid=\"standard\">" << endstr;
+      //mOutput << startstr << "<technique sid=\"standard\">" << endstr;
+      mOutput << startstr << "<technique sid=\"common\">" << endstr;
       PushTag();
       mOutput << startstr << "<phong>" << endstr;
       PushTag();
