Index: scripts/soundplay_node.py
===================================================================
--- scripts/soundplay_node.py   (revision 29682)
+++ scripts/soundplay_node.py   (working copy)
@@ -189,7 +189,7 @@
                         try:
                             txtfile.write(data.arg)
                             txtfile.flush()
-                            os.system('text2wave '+txtfilename+' -o '+wavfilename)
+                            os.system(rospy.get_param('~command','text2wave')+' '+txtfilename+' -o '+wavfilename+' '+rospy.get_param('~command_args',''))
                             try:
                                 if os.stat(wavfilename).st_size == 0:
                                     raise OSError # So we hit the same catch block
@@ -321,7 +321,11 @@
 
     def mixer_init(self): 
         try:
-            mixer.init(11025, -16, 1, 4000)
+            frequency=rospy.get_param('~frequency',11025)
+            size=rospy.get_param('~size',-16)
+            channels=rospy.get_param('~channels',1)
+            buffer=rospy.get_param('~buffer',4000)
+            mixer.init(frequency, size, channels, buffer)
             self.init_vars()
             return True
         except Exception, e:
