diff -r a1bbfb9c4ac3 collada_urdf/CMakeLists.txt
--- a/collada_urdf/CMakeLists.txt	Tue Nov 01 15:26:47 2011 -0700
+++ b/collada_urdf/CMakeLists.txt	Sun Dec 04 09:56:21 2011 +0900
@@ -12,5 +12,11 @@
 rosbuild_add_executable(urdf_to_collada src/urdf_to_collada.cpp)
 target_link_libraries(urdf_to_collada collada_urdf)
 
+include(FindPkgConfig)
+pkg_check_modules(GLIB REQUIRED glib-2.0)
+include_directories(${GLIB_INCLUDE_DIRS})
+link_directories(${GLIB_LIBRARY_DIRS})
+target_link_libraries(collada_urdf gts)
+
 rosbuild_add_gtest(test_collada_writer test/test_collada_urdf.cpp)
 target_link_libraries(test_collada_writer collada_urdf)
diff -r a1bbfb9c4ac3 collada_urdf/manifest.xml
--- a/collada_urdf/manifest.xml	Tue Nov 01 15:26:47 2011 -0700
+++ b/collada_urdf/manifest.xml	Sun Dec 04 09:56:21 2011 +0900
@@ -17,6 +17,8 @@
   <depend package="angles" />
   <depend package="assimp"/>
 
+  <rosdep name="libgts" />
+
   <platform os="ubuntu" version="9.04"/>
   <platform os="ubuntu" version="9.10"/>
   <platform os="ubuntu" version="10.04"/>
