diff -r ee52b9c56e9a collada_urdf/CMakeLists.txt
--- a/collada_urdf/CMakeLists.txt	Thu Jan 10 20:40:58 2013 +0200
+++ b/collada_urdf/CMakeLists.txt	Mon Feb 04 22:29:54 2013 +0900
@@ -47,12 +47,17 @@
   link_directories(${COLLADA_DOM_LIBRARY_DIRS})
 endif()
 
+include(FindPkgConfig)
+pkg_check_modules(GLIB REQUIRED glib-2.0)
+include_directories(${GLIB_INCLUDE_DIRS})
+link_directories(${GLIB_LIBRARY_DIRS})
+
 include_directories(${catkin_INCLUDE_DIRS})
 link_directories(${catkin_LIBRARY_DIRS})
 
 add_library(${PROJECT_NAME} src/collada_urdf.cpp)
 target_link_libraries(${PROJECT_NAME} ${tinyxml_library} ${ASSIMP_LIBRARIES} ${catkin_LIBRARIES} ${COLLADA_DOM_LIBRARIES}  
-  ${Boost_LIBRARIES})
+  ${Boost_LIBRARIES} gts)
 set_target_properties(${PROJECT_NAME} PROPERTIES COMPILER_FLAGS "${ASSIMP_CXX_FLAGS} ${ASSIMP_CFLAGS_OTHER}")
 set_target_properties(${PROJECT_NAME} PROPERTIES LINK_FLAGS "${ASSIMP_LINK_FLAGS}") 
 
