Index: src/collada_writer.cpp
===================================================================
--- src/collada_writer.cpp	(リビジョン 31159)
+++ src/collada_writer.cpp	(作業コピー)
@@ -691,6 +691,17 @@
         if (j == geometry_ids_.end())
             continue;
 
+        if ( urdf_link->visual->material_name.empty() ||
+             (urdf_link->visual->material == NULL) ) {
+          ambient.r = ambient.g = ambient.b = 1; ambient.a = 0;
+          diffuse.r = diffuse.g = diffuse.b = 1; diffuse.a = 0;
+        } else {
+          ambient.r = diffuse.r = urdf_link->visual->material->color.r;
+          ambient.g = diffuse.g = urdf_link->visual->material->color.g;
+          ambient.b = diffuse.b = urdf_link->visual->material->color.b;
+          ambient.a = diffuse.a = urdf_link->visual->material->color.a;
+        }
+
         string geometry_id = j->second;
 
         domEffectRef effect = addEffect(geometry_id, ambient, diffuse);
