diff -r 90c214410af7 colladadom/Makefile
--- a/colladadom/Makefile	Tue Mar 06 14:59:02 2012 -0800
+++ b/colladadom/Makefile	Mon Apr 15 23:10:02 2013 +0900
@@ -1,6 +1,6 @@
 all: installed
 TARBALL = build/collada-dom-2.2.zip
-TARBALL_URL = http://pr.willowgarage.com/downloads/collada-dom-2.2.zip
+TARBALL_URL = 'https://sourceforge.net/p/jsk-ros-pkg/code/HEAD/tree/downloads/collada-dom-2.2.zip?format=raw'
 SOURCE_DIR = build/colladadom
 INITIAL_DIR = build/collada-dom
 UNPACK_CMD = unzip
