--- a/collada_urdf/src/collada_urdf.cpp	Fri Oct 28 14:12:50 2011 -0700
+++ b/collada_urdf/src/collada_urdf.cpp	Sun Dec 11 01:36:19 2011 +0900
@@ -1199,6 +1199,7 @@
         switch (geometry->type) {
         case urdf::Geometry::MESH: {
             urdf::Mesh* urdf_mesh = (urdf::Mesh*) geometry.get();
+	    cgeometry->setName(urdf_mesh->filename.c_str());
             _loadMesh(urdf_mesh->filename, cgeometry, urdf_mesh->scale);
             break;
         }
