diff -r eed8b260bfe0 app_manager/src/app_manager/app_manager.py
--- a/app_manager/src/app_manager/app_manager.py	Mon Apr 23 17:42:58 2012 -0700
+++ b/app_manager/src/app_manager/app_manager.py	Tue May 14 01:05:05 2013 -0700
@@ -248,7 +248,10 @@
 
     def handle_stop_app(self, req):
         rospy.loginfo("handle stop app: %s"%(req.name))
-        return self.stop_app(req.name)
+        if req.name == '*':
+            return self.stop_apps()
+        else:
+            return self.stop_app(req.name)
 
     def app_monitor(self):
         while self._launch:
